### PR TITLE
Add default limit to IS List RPCs

### PR DESCRIPTION
**Pull request by ryaplots**
#### Summary

References #7327 

#### Changes

- Add new pagination default
- Add `withLimit` to check if the limit is set
- Use default if `req.Limit` is not set

#### Testing

1. Make a request to get a list of entities with the page limit unset;
2. The request should return 100 results per page.

##### Steps

Run tests

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Testing: The steps/process to test this feature are clearly explained including testing for regressions.
- [ ] Infrastructure: If infrastructural changes (e.g., new RPC, configuration) are needed, a separate issue is created in the infrastructural repositories.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: CHANGELOG.md
Comment (KrishnaIyer) [inline CHANGELOG.md]: `Default page limit ...`

File: pkg/identityserver/config.go
Comment (KrishnaIyer) [inline pkg/identityserver/config.go]: We generally embed this struct into the IS configuration.

File: pkg/identityserver/store/pagination.go
Comment (KrishnaIyer) [inline pkg/identityserver/store/pagination.go]: ` func WithLimit(set, default uint32) uint32` is more concise.

**Review by KrishnaIyer — CHANGES_REQUESTED** [review]
Functionality is good but missing tests (unit and e2e).

File: pkg/identityserver/config.go
Comment (nicholaspcr) [inline pkg/identityserver/config.go]:
```suggestion
// Pagination default values.
```

File: pkg/identityserver/config.go
Comment (nicholaspcr) [reply pkg/identityserver/config.go]: I agree with Krishna that it is probably best to remain with what is generally established.

It doesn't make much of a difference as the only impact of creating its own struct is the increase on the amount of types exported by the `identityserver` package. The benefit being the GoDoc reference, which is quicker to check when trying to use a Config's field while developing something.

**Review by nicholaspcr — COMMENTED** [review]
While the idea of introducing the default pagination value is nice and I support it, I have two comments about the current implementation.

<hr/>

##### 1 - The default value seems to small

The default value seems quite low, being at 10. This will be imposed on the CLI requests as well and I imagine that the change will surprise users who are used to fetching applications and gateways via the CLI.

I'm not saying that we shouldn't do it but rather increase it to be at least 100 and update the CLI `limit` description to inform that the default is `100` and that `0` is not a valid option.

<hr/>

##### 2 - how the default value is set propagated.

I'm not a fan of injecting the default value of pagination in the context. I rather set this value at the store initialization, especially if it is going to be enforced at every list request.

To be more precise I think it would be better to have [functional options](https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md#implementation) in the bunstore initialization:

https://github.com/TheThingsNetwork/lorawan-stack/blob/4203826edb54c7aa32d4e93d86c4808d0f0bdf46/pkg/identityserver/bunstore/store.go#L142-L148
 
<hr/>

I know this deviates a bit from what is written on the issue in which this PR is based on but I think its worth considering having the responsibility of handling the zero values to the store implementation rather than leaving it on the business logic.

**Comment by nicholaspcr** [comment]
> I know this deviates a bit from what is written on the issue in which this PR is based on but I think its worth considering having the responsibility of handling the zero values to the store implementation rather than leaving it on the business logic.

Forgot to tag you @KrishnaIyer

File: pkg/identityserver/store/pagination.go
Comment (ryaplots) [reply pkg/identityserver/store/pagination.go]: `default` is a reserved keyword, but I used `def` then

**Comment by ryaplots** [comment]
I've changed the implementation and moved the default check to the store.
Do I still need unit tests for this? @KrishnaIyer

File: CHANGELOG.md
Comment (nicholaspcr) [inline CHANGELOG.md]:
```suggestion
- Enforce default page limit on IS List RPCs if a value is not provided in the request.
```

File: pkg/identityserver/storetest/api_key_store.go
Comment (nicholaspcr) [inline pkg/identityserver/storetest/api_key_store.go]: By default if the `page` parameter is zero it should be set to `1`, so we might as well do the test with:
```suggestion
		paginateCtx := store.WithPagination(ctx, 0, 0, &total)
```

File: pkg/identityserver/storetest/api_key_store.go
Comment (nicholaspcr) [inline pkg/identityserver/storetest/api_key_store.go]: This test should fail, simply because the `SetPaginationDefaults` function is not being called anywhere, making the default value zero and the results should be 102 items.


The reason is not running is because `storetest` test methods are manually called in the `pkg/identityserver/bunstore/store_test.go`.

https://github.com/TheThingsNetwork/lorawan-stack/blob/4203826edb54c7aa32d4e93d86c4808d0f0bdf46/pkg/identityserver/bunstore/store_test.go#L137-L143

This means that we have to call the newly added test method in this file:
```go
func TestAPIKeyStore(t *testing.T) {
	t.Parallel()
    
	st := storetest.New(t, newTestStore)
	st.TestAPIKeyStoreCRUD(t)
	st.TestAPIKeyStorePagination(t)
    st.TestAPIKeyStorePaginationDefaults(t)
}
```

The `st.<TEST_NAME>` are executed in sequence (no parallelism), meaning that we can call `store.SetPaginationDefaults`  inside of `TestAPIKeyStorePaginationDefaults`

**Review by nicholaspcr — COMMENTED** [review]
I like having the responsibility of defining the default limit value inside the `store` package, that way its agnostic to the bunstore implementation and the default value of `limit` is used every time we call `store.WithPagination` with a limit set to zero.

There shouldn't be any problems with setting the default value once before starting the bunstore instance since all further interactions with the `store` package's internal variable `paginationDefaults` will be read operations only.


<hr/>

I wrote a few comments regarding the tests as I believe they aren't being executed and most likely should yield an error in their current state, but outside of that this looks pretty good to me.

File: pkg/identityserver/storetest/api_key_store.go
Comment (ryaplots) [reply pkg/identityserver/storetest/api_key_store.go]: Good to know! I will fix this.

File: pkg/identityserver/storetest/user_store.go
Comment (nicholaspcr) [inline pkg/identityserver/storetest/user_store.go]: Looks pretty good, I personally would love to see a second call where we set the `page` parameter to be `2` while the limit is still zero. Simulating someone asking for the second page without specifying the limit value.

**Review by nicholaspcr — APPROVED** [review]
This looks good to me.

I left a comment regarding the tests but besides this the PR looks good to me

**Review by KrishnaIyer — APPROVED** [review]
No additional points from me. This can be merged after @nicholaspcr's comments are addressed.

File: pkg/identityserver/bunstore/store_test.go
Comment (nicholaspcr) [inline pkg/identityserver/bunstore/store_test.go]:
```suggestion
	st.TestUserStorePaginationDefaults(t)
```

File: pkg/identityserver/storetest/user_store.go
Comment (nicholaspcr) [inline pkg/identityserver/storetest/user_store.go:514]: Funny enough this test should fail, we have 10 elements with a default page limit of 7. This means that the second page should yield 3 items.

The reason this isn't failing is because the `UserStoreTest` in `bunstore/store_test.go` is calling the `UserSessionStore` test set. I marked the location with a comment.

<hr/>

Additionally it is probably best to add the same test in the other store tests where the pagination test sets: organization, userSession, applications etc.

**Comment by nicholaspcr** [comment]
One last comment which is a `nit`. Squashing fix commits helps if the reviewers wants to look at some specific commit,  the bad side of it is that it forces you to do a `push -f`.

File: pkg/identityserver/bunstore/store_test.go
Comment (nicholaspcr) [inline pkg/identityserver/bunstore/store_test.go]:
```suggestion
	st.TestApplicationStorePaginationDefaults(t)
```

**Review by nicholaspcr — APPROVED** [review]
Looks good to me.

Before merging please fix the wrong test reference in `TestApplicationStore`.

**Pull request merged**
